File: stack/EvaluateReversePolishNotation.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>

using std::string;
using std::vector;
using std::stack;

class Solution {
public:
    int evalRPN_stack(vector<string>& tokens) {
        stack<int> stk;
        int left, right;
        for (auto token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                right = stk.top();
                stk.pop();
                left = stk.top();
                stk.pop();
                if (token == "+")
                    stk.push(left + right);
                else if (token == "-")
                    stk.push(left - right);
                else if (token == "*")
                    stk.push(left * right);
                else
                    stk.push(left / right);
            }
            else {
                stk.push(std::stoi(token));
            }
        }
        return stk.top();
    }

    int evalRPN(vector<string>& tokens) {
        int x, y;
        string token = tokens.back();
        tokens.pop_back();
        if (is_operator(token)) {
            x = evalRPN(tokens);
            y = evalRPN(tokens);
            if (token == "+")
                y += x;
            else if (token == "-")
                y -= x;
            else if (token == "*")
                y *= x;
            else
                y /= x;
        }
        else {
            y = std::stoi(token);
        }
        return y;
    }
    
private:
    bool is_operator(const string& op) {
        return op.size() == 1 && string("+-*/").find(op) != string::npos;
    }
};

int main()
{
    vector<string> t1 {"2", "1", "+", "3", "*"};
    Solution solu;
    std::cout << solu.evalRPN_stack(t1) << "\n";
    std::cout << solu.evalRPN(t1) << "\n";
    return 0;
}

File: stack/EvaluateReversePolishNotation.py
from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        operators = "+-*/"
        token = tokens.pop()
        if token in operators:
            x = self.evalRPN(tokens)
            y = self.evalRPN(tokens)
            if token == '+':
                y += x
            elif token == '-':
                y -= x
            elif token == '*':
                y *= x
            else:
                y = - ((-y) // x) if x * y < 0 else y // x
        else:
            y = int(token)
        return y

    def evalRPN_stack(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if (token == '+' or token == '-' or token == '*' or token == '/'):
                right = stack.pop()
                left = stack.pop()
                if token == '+':
                    stack.append(left + right)
                elif token == '-':
                    stack.append(left - right)
                elif token == '*':
                    stack.append(left * right)
                else:
                    stack.append(- ((-left) // right) if left * right < 0 else left // right )
            else:
                stack.append(int(token))
            # print(stack)
        return stack[-1]


# t = ["2", "1", "+", "3", "*"]
t = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
s = Solution()
print(s.evalRPN(t.copy()))
print(s.evalRPN_stack(t.copy()))
# print(int(12 / (-11)))
